iterators.tex

```diff
@@ -623,31 +650,186 @@
+\range{i}{j} denotes a range. Let $N$ be the smallest number of
+applications of \tcode{++i} necessary to make \tcode{bool(i == j)} be
+\tcode{true}. \tcode{SizedIterator<I>()} is satisfied if and only if:
+
+\begin{itemize}
+\item If $N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{j - i} is well-defined and equals $N$.
+
+\item If $-N$ is representable by \tcode{difference_type_t<I>},
+      then \tcode{i - j} is well-defined and equals $-N$.
+\end{itemize}
+\end{itemdescr}
+
+\pnum
+The \tcode{disable_sized} type predicate provides a mechanism to enable
+use of iterator types that meet the syntactic requirements but do not
+in fact satisfy \tcode{SizedIterator} with the library.
 
 \pnum
-The \tcode{Sentinel} concept defines requirements for a type that
+\enternote A program that instantiates a library template that requires
+an \tcode{Iterator} with such an iterator type \tcode{I} is ill-formed
+with no diagnostic required unless \tcode{disable_sized<I>} evalutes to
+\tcode{true}~(\ref{structure.requirements}). \exitnote
+
+\enternote The \tcode{SizedIterator} concept is satisfied by
+\tcode{RandomAccessIterator}s~(\ref{iterators.random.access}) and by
+counted iterators~(\ref{counted.iterator}).\exitnote
+
+\ednote{This concept also gives us a way to demote the category of \tcode{move_iterator}s to Input
+while retaining the ability of \tcode{move_iterator} pairs to communicate the range's size to
+container constructors.}
+} %% color{newclr}
+
+\rSec2[iterators.sentinel]{Concept Sentinel}
+\pnum
+The \tcode{Sentinel} concept \oldtxt{defines requirements for a type that
 is an abstraction of the past-the-end iterator. Its values can be
-compared to an iterator for equality.
+compared to an iterator for equality.}\newtxt{specifies requirements on
+the types \tcode{S} and \tcode{I} of a sentinel and an iterator that
+denote a range.}
 
 \indexlibrary{\idxcode{Sentinel}}%
-\begin{codeblock}
-  template <class T, class I>
+\begin{itemdecl}
+  template <class @\oldtxt{T}\newtxt{S}@, class I>
   concept bool Sentinel() {
-    return Regular<T>() &&
+    return Regular<@\oldtxt{T}\newtxt{S}@>() &&
       Iterator<I>() &&
-      EqualityComparable<T, I>();
+      EqualityComparable<@\oldtxt{T}\newtxt{S}@, I>();
   }
-\end{codeblock}
+\end{itemdecl}
+
+{\color{newclr}
+\begin{itemdescr}
+Let \tcode{s} and \tcode{t} be values of type \tcode{S},
+and \tcode{i} a value of type \tcode{I} such that \range{i}{s}
+denotes a range. Types \tcode{S} and \tcode{I} satisfy
+\tcode{Sentinel<S, I>()} if and only if:
+
+\begin{itemize}
+\item \tcode{i == s}, \tcode{s == i},\tcode{i != s}, and
+      \tcode{s != i} are well-defined.
+
+\item \tcode{bool(i == s)} if and only if \tcode{i} and
+      \tcode{s} denote the same element.
+
+\item If \range{i}{s} and \range{i}{t} denote the same range,
+      then \tcode{s == t} is well-defined and \tcode{s}
+      equals \tcode{t}.
```

Comment:
I think this sounds like the right approach.

On Wed, Dec 2, 2015, 7:06 PM Casey Carter <notifications@github.com> wrote:

> In iterators.tex
> <https://github.com/ericniebler/stl2/pull/153#discussion_r46496390>:
>
> > +\begin{itemdescr}
> > +Let \tcode{s} and \tcode{t} be values of type \tcode{S},
> > +and \tcode{i} a value of type \tcode{I} such that \range{i}{s}
> > +denotes a range. Types \tcode{S} and \tcode{I} satisfy
> > +\tcode{Sentinel<S, I>()} if and only if:
> > +
> > +\begin{itemize}
> > +\item \tcode{i == s}, \tcode{s == i},\tcode{i != s}, and
> > +      \tcode{s != i} are well-defined.
> > +
> > +\item \tcode{bool(i == s)} if and only if \tcode{i} and
> > +      \tcode{s} denote the same element.
> > +
> > +\item If \range{i}{s} and \range{i}{t} denote the same range,
> > +      then \tcode{s == t} is well-defined and \tcode{s}
> > +      equals \tcode{t}.
>
> Yes, essentially. It's very easy with stateful sentinels to construct a
> range [i,...j,...s1) such that [j,s2) is the same range as [j,s1) but
> [i,s2) is NOT the same range as [i,s1).
>
> I would begin by removing the requirement that sentinels and iterators are
> EqualityComparable, and then build up semantics for i == s that provide
> the properties algorithms need to work. We need:
>
>    - i == s to be equality-preserving (EP).
>    - i != s to be EP and the complement of ==, i.e., i == s iff !(i != s)
>    - s == i iff i == s
>    - s != i iff i != s
>    - i == s, i != s, s == i, and s != i all have the same definition
>    space.
>    - i == s to be well-defined if [i,s) denotes a range.
>
> All but the last are semantics that EC should have and can be presented as
> a simple relaxation for EC to refine. There's nothing here about == for
> sentinels being valid, or what its domain should be, because the algorithms
> simply don't care.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ericniebler/stl2/pull/153/files#r46496390>.
>